docs/use_bls_wallet_dapp.md

```diff
@@ -0,0 +1,109 @@
+# Use BLS Wallet In Your L2 dApp
+
+This guide will show you how to use BLS Wallet in your L2 dApp (Layer 2 decentralized application) so you can utilize multi-action transactions.
+
+## Download, Install, & Setup Quill
+
+[Quill](../extension/) is a protoype browser extension wallet which intergrates [bls-wallet-clients](../contracts/clients/) to communicate with the BLS Wallet smart contracts & transaction aggregator. It supports most of the functionality in [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193).
+
+Currently, we have the contracts deployed to the networks/chains listed [here](https://github.com/web3well/bls-wallet/tree/main/contracts/networks). If your desired networks isn't there, you can use the [Remote Development](./remote_development.md) contract deployment intrsuctions or request a network deploy by [opening an issue](https://github.com/web3well/bls-wallet/issues/new) or [starting a discussion](https://github.com/web3well/bls-wallet/discussions/new).
+
+Below are the instructions for 2 ways you can add Quill to your browser.
+
+### Prebuilt
+
+Go to the [releases page](https://github.com/web3well/bls-wallet/releases) and scroll down to the latest release. In the `Assets` section, download the extension for either Chrome, Firefox, or Opera. To install, simply drag and drop the file into your browser on the extensions page or follow instructions for installing extensions from a file for your browser.
+
+### From Repo
+
+Follow the instructions in either [Local Development](./local_development.md) or [Remote Development](./remote_development.md) to setup this repo and install Quill.
+
+### Setup Quill
+
+Folow the setup instructions when ou first open Quill, and then create a wallet if there is not one there already.
+
+## Connect Your dApp to Quill
+
+Next, connect your dApp to Quill just like you would any other extension wallet.
+
+```typescript
+const provider = providers.Web3Provider(window.ethereum);
+
+await window.ethereum.request({ method: "eth_accounts" });
+```
+
+## Send Multi-Action Transaction
+
+Finally, you can populate & send your multi-action transaction. In the following example, we will do an approve & swap with a DEX (decentralized exchange) in one transaction.
+
+### Check that window.ethereum Supports Multi-Action Transactions
+
+Since any browser extension wallet could be used, make sure that it is Quill before allowing a multi-action transaction.
+
+```typescript
+const areMultiActionTransactionSupported = "isQuill" in window.ethereum && window.ethereum.isQuill;
```

Comment:
💯

Reply:
```suggestion
const areMultiActionTransactionSupported = window.ethereum.isQuill;
```

Reply:
Using the `in` operator doesn't affect the outcome here afaik.

Also it'd be nice to allow this feature detection without testing against Quill specifically. Do you know whether there's anything like that for `window.ethereum` in the wild? This could lead to others supporting multi-action claiming to be Quill so that they support dApps that detect it this way (the userAgent problem).

Reply:
1. `window.ethereum.isQuill` would return `undefined` if the prop doesn't exist, versus `false` with the `in` check. Could also do `!!window.ethereum.isQuill` to ensure a boolean.
2. There are some ongoing discussions in a few wallet chats I'm part of for feature detection/capability listing but I have not seen any specs or more formalized ideas yet. I think for now this ok as it matches what [MetaMask](https://docs.metamask.io/guide/ethereum-provider.html#ethereum-ismetamask) is doing. We could also add an additional prop to ours such as `window.ethereum.areMultiActionTransactionSupported` to be more explicit and granular.

Reply:
Switched to `!!window.ethereum.isQuill`